File: sp/dart/pingpong/core/tdpnet.dart
import 'dart:io';
import 'dart:convert';
import 'dart:async';
import 'packet.dart'; // Assumes the previous Packet/Entity code is here
import 'helper.dart'; // Assumes the previous Packeter code is here

class TDPnet {
  WebSocket? _socket;
  StreamSubscription? _subscription;

  Future<void> connect(String host) async {
    try {
      _socket = await WebSocket.connect(host);
      
      onConnected(DateTime.now().millisecondsSinceEpoch);

      _subscription = _socket!.listen(
        (dynamic message) {
          try {
            // Attempt to parse as JSON first
            final m = jsonDecode(message.toString());
            handleEvent(m['id'], m['data']);
          } catch (err) {
            // If not JSON, treat as raw packet data
            onData(message.toString());
          }
        },
        onError: (err) {
          onError(err);
          onConnectionError(err.toString());
        },
        onDone: () {
          onConnectionError('Connection closed');
        },
        cancelOnError: true,
      );
    } catch (e) {
      onConnectionError(e.toString());
      rethrow;
    }
  }

  void handleEvent(dynamic id, dynamic data) {
    if (data == null) return;
    
    // Using fromSP as defined in previous step
    final packet = Packet.fromSP(data.toString());
    final subject = packet.peering.subject;
    final parts = subject.split(':');
    final category = parts[0];
    final msgId = parts.length > 1 ? parts[1] : id.toString();

    switch (category) {
      case 'request':
        onRequest(msgId, packet);
        break;
      case 'response':
        onResponse(msgId, packet);
        break;
      case 'event':
        onEvent(msgId, packet);
        break;
    }
  }

  void onData(String e) {
    final packet = Packet.fromString(e);
    final subject = packet.peering.subject;
    final parts = subject.split(':');
    final category = parts[0];
    final id = parts.length > 1 ? parts[1] : '';

    switch (category) {
      case 'request':
        onRequest(id, packet);
        break;
      case 'response':
        onResponse(id, packet);
        break;
      case 'event':
        onEvent(id, packet);
        break;
    }
  }

  // Lifecycle hooks (to be overridden)
  void onConnected(int timestamp) {}
  void onError(dynamic err) {}
  void onConnectionError(String msg) {}
  void onRequest(String id, Packet p) {}
  void onResponse(String id, Packet p) {}
  void onEvent(String id, Packet p) {}

  void sendSignal(String data) {
    _socket?.add(data);
  }

  void sendData(String data) {
    _socket?.add(data);
  }

  void request(String from, String to, String id, Map<String, dynamic> data) {
    final packetStr = Packeter.createSignal(
      from: from,
      to: to,
      subject: 'request:$id',
      data: jsonEncode(data),
    );
    sendSignal(packetStr);
  }

  void response(Packet packet, String id, Map<String, dynamic> data) {
    // Swap peering logic
    packet.peering = Peering(
      packet.peering.to, 
      packet.peering.from, 
      'response:$id'
    );
    packet.transformer.data = jsonEncode(data);

    sendData(packet.toString());
  }

  Future<void> close() async {
    await _subscription?.cancel();
    await _socket?.close();
  }
}

File: sp/java/pingpong/core/TDPnet.java
package pingpong.core;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import org.json.JSONObject;
import java.net.URI;
import java.net.URISyntaxException;

public class TDPnet {
    private InternalClient socket;

    public void connect(String host) throws URISyntaxException {
        this.socket = new InternalClient(new URI(host));
        this.socket.connect();
    }

    // Inner class to handle WebSocket events mapping to the JS logic
    private class InternalClient extends WebSocketClient {
        public InternalClient(URI serverUri) {
            super(serverUri);
        }

        @Override
        public void onOpen(ServerHandshake handshakedata) {
            onConnected(System.currentTimeMillis());
        }

        @Override
        public void onMessage(String message) {
            try {
                // Attempt to parse as JSON first
                JSONObject m = new JSONObject(message);
                handleEvent(m.opt("id"), m.optString("data"));
            } catch (Exception err) {
                // If not JSON, treat as raw packet data
                onData(message);
            }
        }

        @Override
        public void onClose(int code, String reason, boolean remote) {
            onConnectionError("Connection closed: " + reason);
        }

        @Override
        public void onError(Exception ex) {
            TDPnet.this.onError(ex);
            onConnectionError(ex.getMessage());
        }
    }

    public void handleEvent(Object id, String data) {
        if (data == null || data.isEmpty()) return;

        Packet packet = Packet.fromSP(data);
        String subject = packet.peering.subject;
        String[] parts = subject.split(":");
        String category = parts[0];
        String msgId = parts.length > 1 ? parts[1] : (id != null ? id.toString() : "");

        switch (category) {
            case "request": onRequest(msgId, packet); break;
            case "response": onResponse(msgId, packet); break;
            case "event": onEvent(msgId, packet); break;
        }
    }

    public void onData(String e) {
        Packet packet = Packet.fromString(e);
        String[] parts = packet.peering.subject.split(":");
        String category = parts[0];
        String id = parts.length > 1 ? parts[1] : "";

        switch (category) {
            case "request": onRequest(id, packet); break;
            case "response": onResponse(id, packet); break;
            case "event": onEvent(id, packet); break;
        }
    }

    // Lifecycle methods to be overridden by the user
    public void onConnected(long timestamp) {}
    public void onError(Exception err) {}
    public void onConnectionError(String msg) {}
    public void onRequest(String id, Packet p) {}
    public void onResponse(String id, Packet p) {}
    public void onEvent(String id, Packet p) {}

    public void sendSignal(String data) {
        if (socket != null && socket.isOpen()) {
            socket.send(data);
        }
    }

    public void sendData(String data) {
        if (socket != null && socket.isOpen()) {
            socket.send(data);
        }
    }

    public void request(String from, String to, String id, JSONObject data) {
        String packetStr = Packeter.createSignal(
            from,
            to,
            data.toString(),
            "request:" + id,
            null, // default operators
            0     // default pointer
        );
        sendSignal(packetStr);
    }

    public void response(Packet packet, String id, JSONObject data) {
        // Swap peering for response
        packet.peering = new Peering(
            packet.peering.to, 
            packet.peering.from, 
            "response:" + id
        );
        packet.transformer.data = data.toString();

        sendData(packet.toString());
    }
}